### DISCUSSION TITLE
Will the union syntax be removed or documented?

### DISCUSSION THREAD

**Discussion by danieldaeschle** [vlang/v#7221] [General]
Currently this works:
```v
union Test {
	x int
}

fn main() {
	t := Test{}
}
```

What's the behaviour of it? Will this be removed?

**Comment by spytheman** [comment]
No, the union is a pretty useful construct in some low level situations 
(when you have to look at the same data as if it was different types/ from different perspective).
Why should it be removed?

**Reply by danieldaeschle** [reply]
It's new to me. Didn't know it exists. Is it documented anywhere?
Aren't sum types more ore less the better union?

**Reply by Delta456** [reply]
V Unions are the same as C unions. They are used in `strconv` and `regex`.

**Reply by spytheman** [reply]
It should be documented, not removed.

**Reply by danieldaeschle** [reply]
That's nice to know. I just checked out V's documentation and i couldn't find it documented anywhere. Maybe it would be good to make it this syntax more public.

**Reply by Delta456** [reply]
I agree, but is the current functionality fine?

**Reply by JalonSolov** [reply]
Even COBOL and PL/1 had unions.  The type is about as old as computers.  The concept is as old as math...

**Reply by spytheman** [reply]
Yes, the unions work fine - they are used in some lower level modules as @Delta456 pointed out.

**Reply by spytheman** [reply]
> Aren't sum types more ore less the better union?

You can construct a sumtype equivalent (absent syntax support and compiler checks) from a union type + tag. 

The reverse is not true - a union is a lower level construct than a sumtype. 
One use case that sumtypes do not cover for example is when you need a way to read a block of bytes from a file, then treat this unstructured block of data, as if it was a struct. You can do it using pointer casts, but that is usually more involved and will require 
unsafe {} operations. Using an union, you can just do: 
```v
union UFileHeader { 
    data [32]byte 
    info FileHeader
}
```
... then just use .data while reading/writing, and .info when you need access to the fields in that structure.

**Reply by ntrel** [reply]
> You can do it using pointer casts, but that is usually more involved and will require
unsafe {} operations

Reading/writing union fields now shows a warning outside `unsafe`: #7869.
I think if you only need to convert from one type to one other once, use pointer casts.
But as Rust has union now and for interop with C, it's probably good to support it.

**Reply by dumblob** [reply]
@spytheman 

>Why should it be removed?

Maybe because according to C99 and older C standards the behavior is undefined in certain cases (especially in the case you've showed - namely "look at some data differently").

But if we'll define unions as in Rust, that should be fine. Note though Rust's unions are pretty constrained - and that's a lot of work in *V* as it has to handle paddings differently than C, byte order differeontly than C, etc.
